### ISSUE TITLE
Qubes fails to start/activate systemd `graphical-session.target` target

### ISSUE THREAD

**Issue by adrelanos** [QubesOS/qubes-issues#9576]
### Qubes OS release

R4.2.

### Brief summary

The `graphical-session.target` is missing in Qubes.

This breaks Kicksecure's and Whonix's [msgcollector](https://github.com/Kicksecure/msgcollector) [/usr/lib/systemd/user/msgcollector-gui.service](https://github.com/Kicksecure/msgcollector/blob/master/usr/lib/systemd/user/msgcollector-gui.service).

### Steps to reproduce

```
systemctl --user status graphical-session.target
```

### Expected behavior

Compared to KUbuntu:

```
systemctl --user status graphical-session.target
```

```
● graphical-session.target - Current graphical user session
     Loaded: loaded (/usr/lib/systemd/user/graphical-session.target; static)
     Active: active since Sat 2024-11-09 17:52:48 CST; 2 days ago
       Docs: man:systemd.special(7)
```

### Actual behavior

Qubes Debian Template:

```
systemctl --user status graphical-session.target
```

```
○ graphical-session.target - Current graphical user session
     Loaded: loaded (/usr/lib/systemd/user/graphical-session.target; static)
     Active: inactive (dead)
       Docs: man:systemd.special(7)
```

**Comment by marmarek** [comment]
Interestingly, there is no `graphical-session.target` at all in Fedora (but it is documented in the `systemd.special` man page).

**Comment by marmarek** [comment]
This is going to be a bit tricky, as qubes-gui-agent is a system unit, and graphical-session.target is a user unit. I don't think one can bind them across system/user instances boundary. I guess we need a dummy user unit for this...

**Comment by ArrayBolt3** [comment]
On my Kubuntu system, I see that `/usr/lib/systemd/user/plasma-workspace.target` `Requires` graphical-session.target, `BindsTo` it, and is `Before` it. I can test if making and enabling such a service on R4.3 fixes the issue, and make a pull request if so.

**Comment by marmarek** [comment]
@ArrayBolt3 the thing is, VM has no equivalent of plasma-workspace.target _user_ unit. There is only qubes-gui-agent.service _system_ unit.

**Comment by ArrayBolt3** [comment]
@marmarek Right, but all that needs to be done is making some arbitrary dummy unit like you said, and enabling it (at least I think that will work). I wasn't really coming up with any new ideas, just looking for which settings the dummy unit would need in order to do the right thing.

**Comment by ArrayBolt3** [comment]
The following works for me:

```
[template workstation user ~]% cat /usr/lib/systemd/user/force-graphical-session.service 
[Unit]
Description=get graphical-session.target to work
BindsTo=graphical-session.target
Before=graphical-session.target
Requires=graphical-session.target

[Service]
ExecStart=/usr/bin/true
Type=oneshot
RemainAfterExit=true

[Install]
WantedBy=basic.target
```

Then `systemctl --user enable force-graphical-session.service` and reboot. `graphical-session.target` will be active on boot.

```
[template workstation user ~]% systemctl --user status graphical-session.target
● graphical-session.target - Current graphical user session
     Loaded: loaded (/usr/lib/systemd/user/graphical-session.target; static)
     Active: active since Tue 2024-11-12 21:58:55 UTC; 1min 50s ago
       Docs: man:systemd.special(7)

Nov 12 21:58:55 host systemd[793]: Reached target graphical-session.target - Current graphical user session.
```

**Comment by marmarek** [comment]
Well, even it that "works", it would be wrong thing to do, as `graphical-session.target` is supposed to be active when actual graphical session is active, not just always.

**Comment by ArrayBolt3** [comment]
Yeah, that's a good point. Further complicating matters is that enabling a user service in a template doesn't seem to enable it in a DispVM, and disabling the user service in the template doesn't disable it in the DispVM either. Not sure what's happening there, I'm on Qubes OS R4.3.

**Comment by ArrayBolt3** [comment]
Hmm. The way it works in KDE Plasma appears to be something like this:

* SDDM launches `/usr/bin/startplasma-x11`
* `startplasma-x11` does some setup work, then connects to systemd over D-Bus and instructs it to launch the `plasma-workspace-x11.target` user target
* That then kicks off a whole slew of units, including the aforementioned `plasma-workspace.target`
* `plasma-workspace.target` appears to be what reaches `graphical-session.target`.

So it might be possible for the GUI daemon to do sort of the same thing:

* Launching a qube launches the guest-side qubes-gui-daemon within the qube
* qubes-gui-daemon figures out which user the graphical session is running on (how to do this?)
* qubes-gui-daemon launches some subprocess as the user the graphical session will end up running on (or it forks and changes to the specific user, perhaps)
* The subprocess tells systemd to start graphical-session.target, probably using D-Bus

That wouldn't require a dummy systemd unit at all, though it would require some substantial complexity. The potentially tricky parts are going to be figuring out which user to "drop" to, and then doing the D-Bus connection.

**Comment by ArrayBolt3** [comment]
er, replace gui-daemon with gui-agent, I had them mixed up.

**Comment by marmarek** [comment]
The proper place is to plug it into [`qubes-session`](https://github.com/QubesOS/qubes-gui-agent-linux/blob/main/appvm-scripts/usrbin/qubes-session) somehow - this is what is started under Xorg as the user. But still, not sure how to do it to behave as the documentation says:

>        graphical-session.target
           This target is active whenever any graphical session is running. It
           is used to stop user services which only apply to a graphical (X,
           Wayland, etc.) session when the session is terminated. Such
           services should have "PartOf=graphical-session.target" in their
           [Unit] section. A target for a particular session (e. g.
           gnome-session.target) starts and stops "graphical-session.target"
           with "BindsTo=graphical-session.target".

So, we need to ensure not only it gets started at proper time, but also it gets stopped when graphical session (qubes-session / qubes-gui-agent) are stopped.

**Comment by marmarek** [comment]
Or maybe even one step earlier - plug it into qubes-run-xorg to call the whole xinit etc (see final lines of the script) under user unit. But then, monitoring if the X server is still running will get tricky, `qubes-run-xorg` shouldn't exit before X server exits, and `systemctl --user start` doesn't normally wait for the unit to stop... (maybe `systemctl --user --wait start` could be used?)

**Comment by ArrayBolt3** [comment]
To me it seems like qubes-session is the equivalent of startplasma-x11 as far as sessions are concerned. My initial instinct would be to put it here: https://github.com/QubesOS/qubes-gui-agent-linux/blob/main/appvm-scripts/usrbin/qubes-session#L40-L49 Have a user systemd unit with the needed `BindsTo` that calls `qsvc qubes-gui-agent`, and start that unit using `systemctl` in `qubes-session`. When `qubes-gui-agent` dies, so will the unit (if configured properly), which should take down `graphical-session.target` with it.

**Comment by marmarek** [comment]
If going the latter direction, this would need to be done carefully - all the user processes (including most qrexec services, xdg autostart apps etc) are started as children of qubes-session, which means it's sensitive for changes in environment vars, applied limits, cgroups etc...



> My initial instinct would be to put it here: https://github.com/QubesOS/qubes-gui-agent-linux/blob/main/appvm-scripts/usrbin/qubes-session#L40-L49

Nope, this part applies only to sys-gui (it starts Xfce or whatever user has configured in sys-gui, which itself runs gui-agent connected to dom0).

**Comment by ArrayBolt3** [comment]
> Nope, this part applies only to sys-gui (it starts Xfce or whatever user has configured in sys-gui, which itself runs gui-agent connected to dom0).

hmm, I may be severely confused as to the purpose of the qubes-session script then, because I don't see anything else in that script that would launch the GUI agent, and I thought this was the GUI agent launcher each qube ran.

**Comment by marmarek** [comment]
the flow is: qubes-gui-agent -> qubes-run-xorg -> xinit -> ... -> Xorg + qubes-session

**Comment by ArrayBolt3** [comment]
Currently, Whonix 18 is working around this with a file in `/etcc/xdg/autostart` that runs `systemctl --user start msgcollector-gui.service`. This works, but it also seems to be causing the service to be started shortly *after* shutdown has been started if starting and then immediately shutting down a qube. This appears to possibly be causing hangs during shutdown, as shown by the guest-test-inst-dvm.log file from system_tests-var_log.tar.gz at https://openqa.qubes-os.org/tests/162679/#downloads.

**Comment by marmarek** [comment]
Shouldn't service started late still be stopped as part of shutdown?
Anyway, at least in some cases, msgcollector-gui is started before shutdown is initiated and it still fails to get stopped, so that's not the whole story...

**Comment by ArrayBolt3** [comment]
> Shouldn't service started late still be stopped as part of shutdown?

It should be, but it isn't, why is unclear. Maybe the point at which shutdown "starts" comes after the user sessions are ended, and so the fact that the unit starts after the user session ends somehow ends up with it in limbo?

The unit itself is https://github.com/Kicksecure/msgcollector/blob/master/usr/lib/systemd/user/msgcollector-gui.service, the script that starts it is https://github.com/Kicksecure/msgcollector/blob/master/usr/libexec/msgcollector/msgdispatcher_xdg_autostart and the desktop file that kicks off that script is https://github.com/Kicksecure/msgcollector/blob/master/etc/xdg/autostart/msgdispatcher.desktop. The only thing I can see here that's suspicious other than the mechanism itself is the use of `--no-block` in the start command, and I can't really see how this would cause that... msgcollector-gui.service does indeed finish starting, systemd just never tries to stop it (no `Stopping` message is shown for it after it starts in at least some of the failed shutdowns).

**Comment by marmarek** [comment]
Re-posting from Matrix (but it's getting offtopic to `graphical-session.target` thing):
I just noticed something interesting just after the lines I pasted above:
```
[2025-12-18 13:39:18] [   16.280719] systemd-logind[1119]:  [0;1;39mThe system will power off now! [0m
[2025-12-18 13:39:18] [   16.280900] systemd-logind[1119]:  [0;1;39mSystem is powering down. [0m
[2025-12-18 13:39:18] [   16.289117] privleapd[1057]: send_action_results: INFO: Action 'tor-consensus-valid-until' requested by account 'sdwdate' completed
[2025-12-18 13:39:18] [   16.289255] systemd[1]:  [0;1;38:5:185mbasic.target: Found ordering cycle on user@1000.service/stop [0m
[2025-12-18 13:39:18] [   16.289382] systemd[1]:  [0;1;38:5:185mbasic.target: Found dependency on sdwdate-gui-qubes@1000.socket/stop [0m
[2025-12-18 13:39:18] [   16.289491] systemd[1]:  [0;1;38:5:185mbasic.target: Found dependency on sockets.target/stop [0m
[2025-12-18 13:39:18] [   16.289612] systemd[1]:  [0;1;38:5:185mbasic.target: Found dependency on basic.target/stop [0m
[2025-12-18 13:39:18] [   16.289754] systemd[1]:  [0;1;31mbasic.target: Job user@1000.service/stop deleted to break ordering cycle starting with basic.target/stop [0m
[2025-12-18 13:39:18] [   16.289856] systemd[1]:  [0;1;38:5:185mbasic.target: Found ordering cycle on sdwdate-gui-qubes-proxy-helper.service/stop [0m
[2025-12-18 13:39:18] [   16.290016] systemd[1]:  [0;1;38:5:185mbasic.target: Found dependency on sdwdate-gui-qubes@1000.socket/stop [0m
[2025-12-18 13:39:18] [   16.291768] systemd[1]:  [0;1;38:5:185mbasic.target: Found dependency on sockets.target/stop [0m
[2025-12-18 13:39:18] [   16.291922] systemd[1]:  [0;1;38:5:185mbasic.target: Found dependency on basic.target/stop [0m
[2025-12-18 13:39:18] [   16.292821] systemd[1]:  [0;1;31mbasic.target: Job sdwdate-gui-qubes-proxy-helper.service/stop deleted to break ordering cycle starting with basic.target/stop [0m
[2025-12-18 13:39:18] [   16.292963] systemd[1]:  [0;1;38:5:185msockets.target: Found ordering cycle on sdwdate-gui-qubes@1000.socket/stop [0m
[2025-12-18 13:39:18] [   16.294171] systemd[1]:  [0;1;38:5:185msockets.target: Found dependency on sockets.target/stop [0m
[2025-12-18 13:39:18] [   16.294299] systemd[1]:  [0;1;31msockets.target: Job sdwdate-gui-qubes@1000.socket/stop deleted to break ordering cycle starting with sockets.target/stop [0m
```
it removes a bunch of stop jobs, so maybe that's why  something doesn't get stopped?
especially `Job user@1000.service/stop deleted` looks to be relevant here...

**Comment by ArrayBolt3** [comment]
@marmarek Have the recent changes to sdwdate-gui been enough to fix the shutdown hang? (This is off-topic to the original bug report, but will help me know how to prioritize this one.)

**Comment by marmarek** [comment]
Yes, I haven't seen this shutdown hang recently (at least in test runs with developers repository enabled).

**Comment by marmarek** [comment]
xdg-desktop-portal.service on Fedora 44 depends on `graphical-session.target` now.

**Comment by ArrayBolt3** [comment]
Affects 4.3.
